CMakeLists.txt

```diff
@@ -77,23 +70,30 @@ string(REGEX REPLACE " /W[0-4]" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
 
 # Duktape
 set(DUKTAPE_SOURCE_ROOT "" CACHE STRING "Root of the Duktape source tree")
-if (NOT DUKTAPE_SOURCE_ROOT)
+
+if (DUKTAPE_SOURCE_ROOT)
+    add_library(duktape ${DUKTAPE_SOURCE_ROOT}/src/duktape.c ${DUKTAPE_SOURCE_ROOT}/src/duktape.h)
+    target_include_directories(
+            duktape PUBLIC
+            "$<BUILD_INTERFACE:${DUKTAPE_SOURCE_ROOT}/src>"
+            "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>")
+
+    if (CMAKE_CXX_COMPILER_ID MATCHES "Clang|GNU")
+        target_compile_options(duktape PRIVATE -w)
+    elseif (CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+        target_compile_options(duktape PRIVATE /w)
+    endif()
+    set(DUKTAPE_INSTALL duktape)
+elseif(DUKTAPE_PREBUILT)
```

Comment:
What does DUKTAPE_PREBUILT mean?

Reply:
it means I have it installed on my machine, in my case with `dnf install duktape-devel`.

Reply:
OK. I would say we have to document this but https://github.com/cppalliance/mrdox/pull/434, which includes the basic docs, is now hanging.

What is it you get with `dnf install duktape-devel`? Is it equivalent with what you would get from other package manager?

I'm still not sure I understand the logic. Maybe it's some pattern I don't know about. I understand

```
set_target_properties(duktape PROPERTIES IMPORTED_LOCATION ${DUKTAPE_PREBUILT})
```

but then

```
add_library(duktape SHARED IMPORTED src/test-plugin/test-plugin.cpp)
```

because it doesn't have anything to do with duktape and doesn't explain where the headers would come from. How does this become something equivalent to the path where we use DUKTAPE_SOURCE_DIR?

Reply:
Well it's importing the lib. So I have an imported lib instead of a built one, so I need to set the import_location.

I am not saying that this is a great solution, nor that it needs to be in this PR. I can remove it and make a separate one if you like me to

Reply:
> Well it's importing the lib. So I have an imported lib instead of a built one, so I need to set the import_location.

Yes. This is precisely the part I mentioned I understand perfectly:

```
set_target_properties(duktape PROPERTIES IMPORTED_LOCATION ${DUKTAPE_PREBUILT})
```

It's the rest I don't understand. I don't understand is: 

1) how `src/test-plugin/test-plugin.cpp` is being imported since it's not compiled yet and I thought CMake didn't accept in source files in imported targets, 
2) what `src/test-plugin/test-plugin.cpp` semantically has to do with duktape as a target, 
3) how this imported library compares to the libraries we can get from other sources, 
4) how this can be equivalent with the alternative path based on DUKTAPE_SOURCE_DIR that doesn't include any of that and sets other properties 
5) where the headers come from, 
6) how can we achieve the same with the DUKTAPE_SOURCE_DIR pattern everyone else has been using since it doesn't include this file. 

Basically, all the semantics of what you're trying to achieve are unclear to me. I'm not saying it's wrong. I'm literally saying I don't understand it. But leaving modesty aside for a minute, if I don't understand it most people won't understand it either.

Reply:
1. That's a mistake.
2. Mistake.
3. No idea.
4. Because the includes are /usr/include
5. `/usr/include` in my case
6. Can't, because DUKTAPE_SOURCE_DIR is building.

Reply:
I think I should remvoe this - can't we move duktape to FetchContent though?

Reply:
> That's a mistake. Mistake. No idea. Because the includes are /usr/include /usr/include in my case Can't, because DUKTAPE_SOURCE_DIR is building.

Oh... that's pretty bad then. Good thing I asked.

> I think I should remvoe this - can't we move duktape to FetchContent though?

I wrote FetchContent for fmt and duktape but Vinnie hated it. I think his internet was down and he deleted the build dir for some reason then didn't have duktape available. I'm always OK with fetchcontent as a second option, since the DUKTAPE_SOURCE_DIR option was still there, nothing changed in that sense, and I guess I would just tether some internet from my phone for a few seconds, download duktape and get back to it. 

But anyway... I don't think FetchContent helps much in this case. I'll just get the same content you already get from DUKTAPE_SOURCE_DIR. The instructions are in that other PR https://github.com/cppalliance/mrdox/pull/434 I mentioned before.

Reply:
So should I remove this or do you think it's an ok option for linux builds?

Reply:
I'm using Linux most of the time and I don't ever see myself using that. I don't think the problem here is related to the OS at all. The problem is the pattern is not generalizable (although I understand the default hint paths cmake uses on linux do make a difference). `DUKTAPE_PREBUILT` could just be the usual `DUKTAPE_ROOT` cmake uses for all packages, if we had reliable packages. Then we'd have to agree on what these packages are, which is hard (because there are no reliable packages even from the original repo let alone CMake integration for them) but doable and fine. 

The problem at this point is not even about the packages anymore, which is a problem (`src/test-plugin/test-plugin.cpp` is not part of duktape, the duktape target should have the headers, if the library is prebuilt then cmake should be the one finding it, etc...) but that could easily be in another PR. The problem is the thing doesn't work the same when it takes the usual `DUKTAPE_SOURCE_ROOT` path everyone else is using.

I'd leave only `DUKTAPE_SOURCE_ROOT` here for now and if we want to have an alternative path that uses an installed version of duktape, then we'd choose which package that would be, document what it is, make the binary version the default in CML, fall back to `DUKTAPE_SOURCE_ROOT` ensuring both behave the same, and make the script look good by using the usual CMake conventions (the hint path is `DUKTAPE_ROOT` and then we use `find_package` or at least `find_library`/`find_header`). That could be on the same PR or a different one.

Reply:
What if I can make it work with a `DUKTAPE_ROOT` where it checks for source and for a lib, and just picks the right thing?

Reply:
I guess that works. Just notice that `DUKTAPE_ROOT` is already the default cmake uses for the hint if I'm not wrong. If not, it's just a called to `find_*` using it as the hint. So you just really need to `find_library` and `find_path` (is it `find_path` the best command for finding headers(?)). We can leave only `DUKTAPE_SOURCE_ROOT` in the docs for now because `DUKTAPE_ROOT` is kind of unstable in terms of recommending to the user how to install it (also because the PR with these docs is not merged yet).